wasm-text.md

```diff
@@ -360,6 +360,229 @@ Imports can be declared like regular functions, memories, etc., by giving an inl
+
+The text format is one of the concrete formats of Wasm.
+Every concrete format maps to a common structure, described as an abstract syntax.
+The function `text2abstract` is a partial function which maps valid programs in the text format to the abstract format.
+Some classes of invalid programs, such as those where an identifier appears in a context in which it is not declared, are undefined.
+
+The function deals with the desugarings which are context dependent.
+Other desugarings are either left for runtime or expressed as macros (for now).
+
+TODO:
+
+* Get rid of inline type declarations (only allow types defined first, inline type declarations serve as documentation and identifier bindings). Something like `(func (type X) TDS:TDecls ... ) => (func (type X))` and `(func TDS:TDecls ...) => (type TDECLS) (func (type NEXT_TYPE_ID) or something)`
+* Remove module names.
+* Give the text format and core format different types, and have the preprocessing handle the conversion. So that identifiers don't even exist in the core type.
+
+### The Context
+
+The `Context` contains information of how to map text-level identifiers to corresponding indices.
+
+```k
+    syntax Context    ::= ctx(localIds: Map)
+ // ----------------------------------------
+```
+
+### Traversing the Text Format
+
+The program is traversed in full once, context beting gathered along the way.
+Since we do not have polymorphic functions available, we define one function per sort of syntactic construct we need to traverse, and for each type of list we encounter.
+
+```k
+    syntax Stmt      ::= "#t2aStmt"       "<" Context ">" "(" Stmt       ")" [function]
+    syntax Defn      ::= "#t2aDefn"       "<" Context ">" "(" Defn       ")" [function]
+    syntax FuncSpec  ::= "#t2aFuncSpec"   "<" Context ">" "(" FuncSpec   ")" [function]
+    syntax TypeUse   ::= "#t2aTypeUse"    "<" Context ">" "(" TypeUse    ")" [function]
+    syntax LocalDecl ::= "#t2aLocalDecl"  "<" Context ">" "(" LocalDecl  ")" [function]
+ // -----------------------------------------------------------------------------------
+    rule text2abstract(SS) => #t2aStmts<ctx( ... localIds: .Map)>(SS)
+
+    rule #t2aStmt<C>(( module OID:OptionalId DS )) => ( module OID #t2aDefns<C>(DS) )
+    rule #t2aStmt<C>(D:Defn)  => #t2aDefn<C>(D)
+    rule #t2aStmt<C>(I:Instr) => #t2aInstr<C>(I)
+    rule #t2aStmt<_>(S) => S [owise]
+
+    rule #t2aDefn<C>(( func OID:OptionalId FS:FuncSpec )) => ( func OID #t2aFuncSpec<C>(FS))
+    rule #t2aDefn<C>(D:Defn) => D [owise]
+
+    rule #t2aFuncSpec<C> (( export WS ) FS:FuncSpec ) => ( export WS ) #t2aFuncSpec<C>(FS)
+    rule #t2aFuncSpec<C>(T:TypeUse LS:LocalDecls IS:Instrs) => #t2aFuncSpec<C>(#updating(T LS IS))
+    rule #t2aFuncSpec<ctx(... localIds: (_ => #ids2Idxs(T, LS)))>(#updating(T:TypeUse LS:LocalDecls IS:Instrs) => #ready(T LS IS))
+    rule #t2aFuncSpec<C>(#ready(T:TypeUse LS:LocalDecls IS:Instrs)) => #t2aTypeUse<C>(T) #t2aLocalDecls<C>(LS) #t2aInstrs<C>(IS)
```

Comment:
This whole `#updating` and `#ready` thing is confusing. Does it not work with something like:

```
rule #t2aFuncSpec<C>(T:TypeUse LS:LocalDecls IS:Instrs) => #t2aTypeUse<#ids2Idxs(T, LS)> #t2aLocalDecls<#ids2Idxs(T, LS)>(LS) #t2aInstrs<#ids2Idxs(T, LS)>(IS)
```

Reply:
That would work now, but the context is a record because it's supposed to contain more maps. We would want to only mention relevant parts of that context when updating it. I could use your suggestion for now, but it would have to be changed as soon as we expand the context.

The `#fun(ctx(...))(C)` doesn't work either due to a parsing issue that isn't solved yet.

How about making a context-updating function with a relevant name? Like
```
syntax Context ::= #updateLocalIds(Context, Map) [function, functional] | #updateLocalIds(Context, Map, Status) [function, functional]
syntax Status   ::= "#updating" | "#done"

rule #updateLocalIds(C, M) => #updateLocalids(C, M, #updating)
rule #updateLocalIds(ctx(... localIds: (_ => M)), M, #updating => #ready)
rule #updateLocalIds(C, _, #ready) => C
```

```
rule #t2aFuncSpec<C>(T:TypeUse LS:LocalDecls IS:Instrs)
  => #t2aTypeUse   <#updateLocalIds(C, #ids2Idxs(T, LS))>(T)
     #t2aLocalDecls<#updateLocalIds(C, #ids2Idxs(T, LS))>(LS)
     #t2aInstrs    <#updateLocalIds(C, #ids2Idxs(T, LS))>(IS)
```

It's not pretty, but at least it hides the complexity under a relevant name.

Reply:
Or we could just skip using the record and use a map from `String` to `Map`. In the end, the context will contain at least 7 maps (locals, labels, functions, types, globals, memories, tables) and one integer (label depth) if we want to get rid of `#ContextLookup`, so always mentioning the whole context would become ugly fast.

Reply:
The context-updating function seems like the best choice right now to me, because it allows us to use the nice record type.

Just define it at the same place the `Context` record is defined:

```
syntax Context ::= ctx ( localIds: Map )
            | #updateLocalIds ( Context , Map ) [function]
            | #updateLocalIdsAux ( Context , Map , Status ) [function]
// ------------------------
// Definition of `#updateLocalIds` and `#updateLocalIdsAux` here.
```

Until we have record-update syntax, this will have to do.